src/liboslexec/llvm_ops.cpp

```diff
@@ -148,9 +148,9 @@ void * __dso_handle = 0; // necessary to avoid linkage issues in bitcode
 
 #ifndef OSL_SHADEOP
 #  ifdef __CUDACC__
-#    define OSL_SHADEOP extern "C" __device__ OSL_LLVM_EXPORT
+#    define OSL_SHADEOP extern "C" __device__ OSL_LLVM_EXPORT __attribute__((always_inline))
```

Comment:
How about `OSL_FORCEINLINE` instead? It's already got multi-platform/compiler logic about how to express the attribute.

Reply:
`OSL_FORCEINLINE` includes `inline` as well which changes the linking semantics and causes a cascade of other errors.

Since this file is only compiled by `clang`, it seemed safe to rely on their syntax directly.

Reply:
OK, N/M

What happens with inline and __attribue__((always_inline)) are combined? does that mean something different?

Reply:
`inline` in C++ has more to do with changing the linking rules. I don't think there is anything special about the combination, its just that in this particular context, only the "stronger" variant is needed.

In regular C++ code you would probably rarely want to use one without the other because the `always_inline` keyword would typically apply to functions whose body is in a header file and you'd want to let the linker know that its ok to see multiple copies of it. The `always_inline` keyword is more of a hint directly to the guts of the optimizer, without any special linker semantics.

Reply:
I see. All good then.